haystack/preview/components/builders/prompt_builder.py

```diff
@@ -1,41 +1,136 @@
+    :param template_variables: (Optional) A list of all template variables to be used as input.
+                              This parameter enables dynamic templating based on user messages. Defaults to None.
+
+    :raises ValueError: If neither `template` nor `template_variables` are provided.
+
+
+    Usage (static templating):
 
-    Usage:
     ```python
     template = "Translate the following context to {{ target_language }}. Context: {{ snippet }}; Translation:"
     builder = PromptBuilder(template=template)
     builder.run(target_language="spanish", snippet="I can't speak spanish.")
     ```
+
+    The above template is used for all messages that are passed to the `PromptBuilder`.
+
+
+    Usage (dynamic templating):
+    ```python
+
+    from haystack.preview.dataclasses.chat_message import ChatMessage
+    template = "What's the weather like in {{ location }}?"
+
+    prompt_builder = PromptBuilder(template_variables=["location", "time"])
+    messages = [ChatMessage.from_system("Always start response to user with Herr Blagojevic.
+    Respond in German even if some input data is in other languages"),
+                ChatMessage.from_user(template)]
+
+    response = pipe.run(data={"prompt_builder": {"location": location, "messages": messages}})
+    ```
+
+    In this example, only the last user message is templated. The template_variables parameter in the PromptBuilder
+    initialization specifies all potential template variables, yet only the variables utilized in the template are
+    required in the run method. For instance, since the time variable isn't used in the template, it's not
+    necessary in the run method invocation above.
+
+
+    Note:
+        The behavior of `PromptBuilder` is determined by the initialization parameters. A static template
+        provides a consistent prompt structure, while template variables offer dynamic templating per user
+        message.
+
     """
 
-    def __init__(self, template: str):
+    def __init__(self, template: Optional[str] = None, template_variables: Optional[List[str]] = None):
         """
-        Initialize the component with a template string.
+        Initialize the component with either a template string or template variables.
+        If template is given PromptBuilder will parse the template string and use the template variables
+        as input types. Conversely, if template_variables are given, PromptBuilder will directly use
+        them as input variables.
+
+        If neither template nor template_variables are provided, an error will be raised.
 
-        :param template: Jinja2 template string, e.g. "Summarize this document: {documents}\nSummary:"
-        :type template: str
+        :param template: Template string to be rendered.
+        :param template_variables: List of template variables to be used as input types.
         """
+        # dynamic per-user message templating
+        if template_variables:
```

Comment:
I think we should add a check to verify that either `template` or `template_variables` are used.

As of now I can write this and it doesn't fail, I don't think it's intended? 🤔 
```
PromptBuilder(template="Foo {{bar}}", template_variables=["test"])
```